pkg/controllers/provisioning/scheduling/topologydomaingroup.go

```diff
@@ -0,0 +1,67 @@
+/*
+Copyright The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scheduling
+
+import (
+	v1 "k8s.io/api/core/v1"
+
+	"sigs.k8s.io/karpenter/pkg/scheduling"
+)
+
+// TopologyDomainGroup tracks the domains for a single topology. Additionally, it tracks the taints associated with
+// each of these domains. This enables us to determine which domains should be considered by a pod if its
+// NodeTaintPolicy is honor.
+type TopologyDomainGroup map[string][][]v1.Taint
```

Comment:
Does this always work? I'm concocting a scenario where there is a single node in a domain (let's say zone) that has a pod scheduled to it and a taint and another node in a domain (same zone) that has a pod scheduled to it that matches the nodeSelector. This code would mean that we would count both pods when determining domain selection -- since there is one set of taints that we tolerate in the domain. Do I have that right?

Reply:
If I understand correctly, I believe I have a [test](https://github.com/kubernetes-sigs/karpenter/blob/01de6ac7643b419957839b1a66e9aa33ab6856c6/pkg/controllers/provisioning/scheduling/topology_test.go#L1459-L1525) for this exact scenario. 

The role of the `TopologyDomainGroup` is solely to determine the eligible domains when constructing a `TopologyGroup`. A domain is eligible if the pod tolerates any of the groups of taints stored in the domain group. This isn't actually responsible for determining which nodes are counted as part of a domain, that's still going to be handled by the `TopologyNodeFilter`, where we still verify that the node is tolerated by the pod for which the `TopologyGroup` was constructed.